onnxruntime/core/framework/utils.cc

```diff
@@ -342,12 +342,19 @@ static bool FinalizeCopyInfoForFetches(gsl::span<const OrtMemoryInfo* const>& fe
   for (size_t i = 0; i < num_outputs; ++i) {
     const OrtMemoryInfo* alloc_info = fetch_alloc_info[i];
 
-    if (alloc_info != nullptr) {
+    if (alloc_info) {
+      // Somehow we get a location requirement for this fetched OrtValue,
+      // let's honor it.
       copy_info[i].target_device = alloc_info->device;
```

Comment:
IOBinding allows the user to specify the device even if the shape is unknown ahead of time so the output cannot be pre-allocated.

Reply:
IOBinding is super slow. I have a few hundred inputs + outputs, and IOBinding time is as long as the actual compute time.